fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java

```diff
@@ -153,14 +159,191 @@ public OptExpression optimize(ConnectContext connectContext,
+
+            if (planCountOfKProperties + entry.getValue() >= localRankOfGroupExpression) {
+                // 4. compute the localProperty-rank in the property.
+                int localRankOfProperties = localRankOfGroupExpression - planCountOfKProperties;
+                // 5. compute sub-ranks of children groups
+                List<Integer> childRankList =
+                        computeNthOfChildGroups(chooseGroupExpression, localRankOfProperties, inputProperties);
+                // 6. computes the child group recursively
+                for (int childIndex = 0; childIndex < chooseGroupExpression.arity(); ++childIndex) {
+                    OptExpression childPlan = extractNthPlanImpl(inputProperties.get(childIndex),
+                            chooseGroupExpression.inputAt(childIndex), childRankList.get(childIndex));
+                    childPlans.add(childPlan);
+                }
+                break;
+            }
+            planCountOfKProperties += entry.getValue();
+        }
+        // 7. construct the OptExpression
+        OptExpression chooseExpression = OptExpression.create(chooseGroupExpression.getOp(), childPlans);
+        chooseExpression.setStatistics(group.getConfidenceStatistics() != null ?
+                group.getConfidenceStatistics() :
+                group.getStatistics());
+
+        // When build plan fragment, we need the output column of logical property
+        chooseExpression.setLogicalProperty(group.getLogicalProperty());
+        return chooseExpression;
+    }
+
+    // compute sub-ranks of children groups
+    private List<Integer> computeNthOfChildGroups(GroupExpression groupExpression, int localNth,
+                                                  List<PhysicalPropertySet> inputProperties) {
+        // 1. compute the plan count of each child group which satisfied the inputProperties
+        List<Integer> childGroupExprCountList = Lists.newArrayList();
+        for (int childIndex = 0; childIndex < groupExpression.arity(); ++childIndex) {
+            PhysicalPropertySet childInputProperty = inputProperties.get(childIndex);
+            Set<GroupExpression> childGroupExpressions =
+                    groupExpression.inputAt(childIndex).getSatisfyRequiredGroupExpressions(childInputProperty);
+            int childGroupExprPropertiesCount = childGroupExpressions.stream()
+                    .mapToInt(childGroupExpression -> childGroupExpression
+                            .getRequiredPropertyPlanCount(childInputProperty)).sum();
+            childGroupExprCountList.add(childGroupExprPropertiesCount);
+        }
+
+        // 2. use the childGroupExprPropertiesCount to compute nth rank of each child group
+        List<Integer> childNthList = Lists.newArrayList();
+        // use this variable as R(i+1) to avoid the recursive calculation.
+        AtomicInteger subsequentChildGroupRank = new AtomicInteger();
```

Comment:
It's will be multi-thread？

Reply:
No，just want to modify subsequentChildGroupRank in function